### ISSUE TITLE
Build TriBITS Documentation with Sphinx and deploy to the web

### ISSUE THREAD

**Issue by bartlettroscoe** [TriBITSPub/TriBITS#386]
CC: @marcinwrobel1986, @fnrizzi 

## Description

As discussed at the prior meeting on 6/24/2021, this Issue is to investigate the building and rendering of TriBITS documentation written in reStructuredText using Sphinx and deploy somehow to the web.

There are three documents that are generated using as reStructuredText files using the script:

```
$ cd TriBITS/
$ cd tribits/doc/
$ ./build_docs.sh
```

All the details of what gets built and how can be found in that script and the scripts and tools it calls.

To build the documentation locally, one needs to have docutils installed.  But even without docutils installed, these scripts will still build everything needed to for the assembled files:

```
tribits/doc/guides/maintainers_guide/TribitsMaintainersGuide.rst
tribits/doc/guides/users_guide/TribitsUsersGuide.rst
tribits/doc/build_ref/TribitsBuildReference.rst
```

The script:

```
tribits/doc/publish_docs.sh
```

is used to publish the generated *html files on the tribits.org site (but you can run it locally to test it out).

## Tasks

* [x] Implement sphinx build of the documents ... See PR #387 
* [x] Deploy sphinx documents as a GitHub pages site ... See the site https://tribitspub.github.io/TriBITS/ served off of of the branch https://github.com/TriBITSPub/TriBITS/tree/deploy-doc-site
* [x] Update tribits.org to point  to new site and new in-progress Sphinx-generated documents
* [ ] Create follow-on GitHub Issue listing remaining issues that need to be resolved before abandoning the DocUtils-generated documentation and fully adopting the Sphinx-generated documentation.

**Comment by bartlettroscoe** [comment]
CC: @fnrizzi, @keitat 

@marcinwrobel1986, as we discussed at last Thursday's meeting, this is the issue to investigate using Sphinx and readthedocs.com to improve the rendering of the TriBITS documentation currently hosted at tribits.org.

**Comment by marcinwrobel1986** [comment]
please find a new documentation under this address:
https://marcinwrobel1986.github.io/TriBITS/

What was done:
- a python script for combining together documentation for Sphinx
- Sphinx configuration with the landing page
- CI (github actions) for generating and deploying docs to GitHub pages

Potential threats:
- script `build_docs.sh` needs to be run from it's directory
- it's hard to destinguish any sensible debug method at the current docs generating process
- currently script `build_docs.sh` runs other bash scripts which runs other bash and python scripts
- the desired way would be a single python3 script, which would be easily debuggable as well as safe in terms of language development plans for many years
- currently the biggest concern is a fact that one needs to have python2 on the system which needs to be found under alias `python` (`python2` not working), which consequence is that documentation can't be build when only python3 is available
- DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.

Actions to be taken:
- get rid of python2 dependencies (starting from documentation generation)
- when doing above, refactor scripts making them PEP8 compatible (increase in readability and maintainability)

Regarding where documentation is hosted:
- the easiest way to go would be github pages
- documentation is generated as static page and could be upload to any server
- github pages supports a custom domain as well
- https://readthedocs.org/ could be an option as well, but does not provide anything special and needs additional resources to be setup

**Comment by bartlettroscoe** [comment]
@marcinwrobel1986, thanks for putting this together.    This looks very nice.   Do you have a branch where we can see these changes to make this work?

> script build_docs.sh needs to be run from it's directory

That is easy to fix to allow to be run from anywhere.

> it's hard to destinguish any sensible debug method at the current docs generating process

That is a matter of opinion.  I personally find it much easier to debug smaller independent scripts that call each other.

> currently script build_docs.sh runs other bash scripts which runs other bash and python scripts

That makes them easy to develop and debug (IMHO).

> the desired way would be a single python3 script, which would be easily debuggable as well as safe in terms of language development plans for many years

A single python3 script is harder to develop and debug, IMHO.

> currently the biggest concern is a fact that one needs to have python2 on the system which needs to be found under alias python (python2 not working), which consequence is that documentation can't be build when only python3 is available

Indeed.  The underlying Python tool `extract_rst_cmake_doc.py` that is needed to generate the needed `*.rst` files currently requires Python 2.7.  But it will not be hard to make it work with Python 3.x as well.  Most of the rest of the Python code in TriBITS works with Python 2.7 and 3.x.  See, for example, `gitdist.py`, `SnapshotDir.py`,  `CreateIssueTrackerFromCDashQuery.py`, `GeneralScriptSupport.py`, `CheckinTest.py`, `CDashQueryAnalyzeReport.py` and `Python2and3.py`.  We have to support Python 2.7 as long as RedHat 7 is the primary OS at SNL and there is no Python3 on many systems (only Python 2.7).

It is a bit tricky to make this work smoothly on various systems but it goes like this:

* You use `#!/usr/bin/env python` at the top of the script.
* On systems where only Python 2.7 exists, `python` points to `python2`.  On systems where you want to use Python 3.x, `python` points to `python3`.
* You use conditional code in `Python2and3.py`.

That is it.

We can discuss the rest.

**Comment by bartlettroscoe** [comment]
I found the branch through the commits:

* https://github.com/TriBITSPub/TriBITS/compare/master...marcinwrobel1986:386-Build-TriBITS-Documentation-with-Sphinx-and-ReadTheDocs

**Comment by marcinwrobel1986** [comment]
> I found the branch through the commits:
> 
> * [master...marcinwrobel1986:386-Build-TriBITS-Documentation-with-Sphinx-and-ReadTheDocs](https://github.com/TriBITSPub/TriBITS/compare/master...marcinwrobel1986:386-Build-TriBITS-Documentation-with-Sphinx-and-ReadTheDocs)

That is correct

**Comment by marcinwrobel1986** [comment]
> > script build_docs.sh needs to be run from it's directory
> 
> That is easy to fix to allow to be run from anywhere.

I agree.
> > it's hard to destinguish any sensible debug method at the current docs generating process
> 
> That is a matter of opinion. I personally find it much easier to debug smaller independent scripts that call each other.
> 
> > currently script build_docs.sh runs other bash scripts which runs other bash and python scripts
> 
> That makes them easy to develop and debug (IMHO).
> 
> > the desired way would be a single python3 script, which would be easily debuggable as well as safe in terms of language development plans for many years
> 
> A single python3 script is harder to develop and debug, IMHO.

When running Python script/scripts with debugger you can stop whenever you like and check the stack/variables, you don't have to print them out like when running separate scripts. IMHO it would make debugging a lot easier.
> > currently the biggest concern is a fact that one needs to have python2 on the system which needs to be found under alias python (python2 not working), which consequence is that documentation can't be build when only python3 is available
> 
> Indeed. The underlying Python tool `extract_rst_cmake_doc.py` that is needed to generate the needed `*.rst` files currently requires Python 2.7. But it will not be hard to make it work with Python 3.x as well. Most of the rest of the Python code in TriBITS works with Python 2.7 and 3.x. See, for example, `gitdist.py`, `SnapshotDir.py`, `CreateIssueTrackerFromCDashQuery.py`, `GeneralScriptSupport.py`, `CheckinTest.py`, `CDashQueryAnalyzeReport.py` and `Python2and3.py`. We have to support Python 2.7 as long as RedHat 7 is the primary OS at SNL and there is no Python3 on many systems (only Python 2.7).
> 
> It is a bit tricky to make this work smoothly on various systems but it goes like this:
> 
> * You use `#!/usr/bin/env python` at the top of the script.
> * On systems where only Python 2.7 exists, `python` points to `python2`.  On systems where you want to use Python 3.x, `python` points to `python3`.
> * You use conditional code in `Python2and3.py`.
> 
> That is it.
> 
> We can discuss the rest.

Separation of documentation creation could be a good idea, because it does not break TriBITS in any way, migration to Python3 will happend in the future anyway and documentation creation can be completely passed to CI (E.g. GitHub actions as TriBITS is hosted on GitHub).

**Comment by bartlettroscoe** [comment]
I am finding a few issues with the documentation at:

* https://marcinwrobel1986.github.io/TriBITS/

For example, it looks like some searches are not finding things like:

* https://marcinwrobel1986.github.io/TriBITS/search.html?q=tribits_add_test&check_keywords=yes&area=default

does not show hits for:

* https://marcinwrobel1986.github.io/TriBITS/users_guide.html#tribits-add-test
* https://marcinwrobel1986.github.io/TriBITS/maintainers_guide.html#tribits-add-test

How do we fix that?

**Comment by bartlettroscoe** [comment]
@marcinwrobel1986, another issue is that bulleted lists are not being maintained like at:

* https://marcinwrobel1986.github.io/TriBITS/maintainers_guide.html#tribits-read-all-project-deps-files-create-deps-graph

it does not show those links to subsections under bullets while they are shown at:

* https://tribits.org/doc/TribitsMaintainersGuide.html#tribits-read-all-project-deps-files-create-deps-graph

Any idea why it is doing that?

Should we just go ahead and create a DRAFT PR for:

* https://github.com/TriBITSPub/TriBITS/compare/master...marcinwrobel1986:386-Build-TriBITS-Documentation-with-Sphinx-and-ReadTheDocs

and do the detailed review there?

**Comment by marcinwrobel1986** [comment]
@bartlettroscoe please point out all the problems Ross and I will take a look at them. As I mentioned I haven't really changed the rst apart from linking.

**Comment by bartlettroscoe** [comment]
@marcinwrobel1986,

Should we just go ahead and create a DRAFT PR for:

* master...marcinwrobel1986:386-Build-TriBITS-Documentation-with-Sphinx-and-ReadTheDocs

and do the detailed review there?

That way, we can reference specific lines of code in *.rst and other files.

**Comment by bartlettroscoe** [comment]
> Separation of documentation creation could be a good idea, because it does not break TriBITS in any way

@marcinwrobel1986, what do you mean by this?

**Comment by bartlettroscoe** [comment]
> When running Python script/scripts with debugger you can stop whenever you like and check the stack/variables, you don't have to print them out like when running separate scripts. IMHO it would make debugging a lot easier.

@marcinwrobel1986, what debugger are you referring to here with Python3?    My past attempts at using 'pdb' for example, were not very positive (but that was a long time ago).

**Comment by marcinwrobel1986** [comment]
> @marcinwrobel1986,
> 
> Should we just go ahead and create a DRAFT PR for:
> 
> * [master...marcinwrobel1986:386-Build-TriBITS-Documentation-with-Sphinx-and-ReadTheDocs](https://github.com/TriBITSPub/TriBITS/compare/master...marcinwrobel1986:386-Build-TriBITS-Documentation-with-Sphinx-and-ReadTheDocs)
> 
> and do the detailed review there?
> 
> That way, we can reference specific lines of code in *.rst and other files.

I agree. I am on it.

**Comment by marcinwrobel1986** [comment]
> > Separation of documentation creation could be a good idea, because it does not break TriBITS in any way
> 
> @marcinwrobel1986, what do you mean by this?

What I mean by that is Python2 is considered to be a technological debt. This debt needs to be paid off sooner or later, right?. You mentioned TriBITS uses Python2. I thaught as now we have resources we could separate documentation to use only Python3 and leave TriBITS itself with Python2. Process of generating documentation will be separated from TriBITS logic, so it will not break anything there.

**Comment by marcinwrobel1986** [comment]
> > When running Python script/scripts with debugger you can stop whenever you like and check the stack/variables, you don't have to print them out like when running separate scripts. IMHO it would make debugging a lot easier.
> 
> @marcinwrobel1986, what debugger are you referring to here with Python3? My past attempts at using 'pdb' for example, were not very positive (but that was a long time ago).

I didn't think about any specific debugger, I was thinking more about developing process which comes with IDE such a PyCharm (tool I am using for Python and web development) or VSC (Visual Studio Code). If one would like to see the stack/variables in any specific place in the code one marks the line, debugger stops there and one can inspect execution of the code.

**Comment by marcinwrobel1986** [comment]
> I am finding a few issues with the documentation at:
> 
> * https://marcinwrobel1986.github.io/TriBITS/
> 
> For example, it looks like some searches are not finding things like:
> 
> * https://marcinwrobel1986.github.io/TriBITS/search.html?q=tribits_add_test&check_keywords=yes&area=default
> 
> does not show hits for:
> 
> * https://marcinwrobel1986.github.io/TriBITS/users_guide.html#tribits-add-test
> * https://marcinwrobel1986.github.io/TriBITS/maintainers_guide.html#tribits-add-test
> 
> How do we fix that?

Hello Ross,

I checked that. After you search you need to choose in which document were you looking for this phrase as documentation is build from `Project, Build, Test, and Install`, `Users Guide` and `Maintainers Guide` please see attached screenshot

![Zrzut ekranu z 2021-07-09 14-18-38](https://user-images.githubusercontent.com/47597569/125077107-00b36980-e0c1-11eb-813d-003cc91ed003.png)

Then after you click on the results, it takes you to the page with phrase highlighted, please see below

![phrase highlighted](https://user-images.githubusercontent.com/47597569/125078324-8552b780-e0c2-11eb-92fa-cb0134659162.png)

And then you can scroll down or use CTRL+F to look for all occurrences

![phrase found](https://user-images.githubusercontent.com/47597569/125078356-90a5e300-e0c2-11eb-9bd8-fef8ff3d1734.png)

I tried to build the documentation, just from one file, because in `Project, Build, Test, and Install`, `Users Guide` and `Maintainers Guide` there are duplicates, but the result was the same.
I will search a bit more and hopefully come back with the answer if there is anything that could be done.

**Comment by bartlettroscoe** [comment]
@marcinwrobel1986, I tried to follow your seatch suggestions [above](https://github.com/TriBITSPub/TriBITS/issues/386#issuecomment-877161915) but I just can't get any type of reasonable search to work.  One advantage of a stupid simple single HTML page like:

* https://tribits.org/doc/TribitsMaintainersGuide.html

is that you just hit "Ctrl+F" with "tribits_add_test" and you jump right to the text that matches.

We need to find a way to fix searches or this Sphinx documentation is unusable, IMHO.

**Comment by bartlettroscoe** [comment]
> I didn't think about any specific debugger, I was thinking more about developing process which comes with IDE such a PyCharm (tool I am using for Python and web development) or VSC (Visual Studio Code).

@marcinwrobel1986, what OS are you using these on?

**Comment by marcinwrobel1986** [comment]
> > I didn't think about any specific debugger, I was thinking more about developing process which comes with IDE such a PyCharm (tool I am using for Python and web development) or VSC (Visual Studio Code).
> 
> @marcinwrobel1986, what OS are you using these on?

I am using Linux, Pop OS distro, ubuntu derivative, but it does not matter, it works with Linux, Windows and MacOS

![Zrzut ekranu z 2021-07-09 16-28-35](https://user-images.githubusercontent.com/47597569/125093652-ce126c80-e0d2-11eb-8e9c-90d32ce06bdb.png)

**Comment by bartlettroscoe** [comment]
> What I mean by that is Python2 is considered to be a technological debt. This debt needs to be paid off sooner or later, right?. You mentioned TriBITS uses Python2. I thaught as now we have resources we could separate documentation to use only Python3 and leave TriBITS itself with Python2. Process of generating documentation will be separated from TriBITS logic, so it will not break anything there.

Right, but as long as RHEL 7 machines have Python 2.7 by default and on Python 3.x, TriBITS Python scripts will continue to support it.  But must Python code in TriBITS also supports Python 3.x as well using the the tools in [Python2and3.py](https://github.com/TriBITSPub/TriBITS/blob/master/tribits/python_utils/Python2and3.py).  We can do this for [extract_rst_cmake_doc.py](https://github.com/TriBITSPub/TriBITS/blob/master/tribits/python_utils/extract_rst_cmake_doc.py) as well.  So all you need is `python` in your path that points to Python 2.7 or Python 3.x and you should be good to go.  I can make that happen.

**Comment by marcinwrobel1986** [comment]
> @marcinwrobel1986, I tried to follow your seatch suggestions [above](https://github.com/TriBITSPub/TriBITS/issues/386#issuecomment-877161915) but I just can't get any type of reasonable search to work. One advantage of a stupid simple single HTML page like:
> 
> * https://tribits.org/doc/TribitsMaintainersGuide.html
> 
> is that you just hit "Ctrl+F" with "tribits_add_test" and you jump right to the text that matches.
> 
> We need to find a way to fix searches or this Sphinx documentation is unusable, IMHO.

I tried a different theme, but the behaviour is the same.

![Zrzut ekranu z 2021-07-09 16-38-15](https://user-images.githubusercontent.com/47597569/125095180-344bbf00-e0d4-11eb-8493-ef2899adc199.png)
![Zrzut ekranu z 2021-07-09 16-38-33](https://user-images.githubusercontent.com/47597569/125095202-39a90980-e0d4-11eb-84ea-fc16bd83a133.png)
![Zrzut ekranu z 2021-07-09 16-38-47](https://user-images.githubusercontent.com/47597569/125095213-3d3c9080-e0d4-11eb-9b41-329349f0e1f6.png)
![Zrzut ekranu z 2021-07-09 16-38-59](https://user-images.githubusercontent.com/47597569/125095222-3f9eea80-e0d4-11eb-94d5-9a97e20e84eb.png)

**Comment by bartlettroscoe** [comment]
> I tried a different theme, but the behaviour is the same.

@marcinwrobel1986, I fear this might turn into a huge time sink.  I boggles the mind how these types of searches don't work.  But the formatting and layout advantages of the Sphinx documentation are large enough that I think it is worth spending some time trying to to address the issues.

In the meantime, even if we merge this branch and set up auto-deployments, it would be good to keep the existing single HTML page versions of this documentation for now.

**Comment by bartlettroscoe** [comment]
@marcinwrobel1986, okay, I see the issue now.  The issue is that each document is a single HTML page which shows up as "introduction" in as title in the search results so it is deceptive what is is showing you.  Once you are on one of the three different document HTML pages, you can do "Ctrl+F".  What throws you off is **Introduction** as the title for the search results and not the title of the document itself.

I appears there is no way to get Sphinx to automatically break up a long RST document into different HTML pages based on section headings with next and previous links a described in:

> https://stackoverflow.com/questions/44777009/can-i-split-a-single-restructuredtext-file-into-multiple-html-pages-by-section

The documentation at:

* https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html

says:

> Since reST does not have facilities to interconnect several documents, or split documents into multiple output files, Sphinx uses a custom directive to add relations between the single files the documentation is made of, as well as tables of contents. The toctree directive is the central element.

So this is a bit problematic.  If you want really nice generated HTML documentation with each section getting its own HTML page, you can't use a single *.rst file and you have to use a Sphinx RST extension directive.  That means if you use this Sphinx directive/extension, then you can no longer use simple docutils to generate HTML and PDF files but have to use Sphinx.  But perhaps that is just the way it is.   (It is just that I have had trouble in the past getting Sphinx to install on machines that I had to use.  For example, I have not tried to install Sphinx on one of my RHEL 7 machines at SNL yet but I can imagine it will be a nightmare like all Python-based installs on RHEL 7.  More on this in PR #387.)

That means that the "Search docs" box is basically worthless.  That is okay if you understand what is happening and that you just need to use "Ctrl+F" to search the HTML page but that was not immediately obvious to me.  (And I realized that I might have mistyped 'tribits_add_test' in my earlier comment [above](https://github.com/TriBITSPub/TriBITS/issues/386#issuecomment-877225582) and that is why it did not hit anything with "Ctrl+F".)

**Comment by bartlettroscoe** [comment]
FYI: I am setting up a meeting between myself and @marcinwrobel1986 (tentatively on Tuesday) to discuss above and related topics.

----

From: Bartlett, Roscoe A 
Sent: Sunday, July 11, 2021 6:46 PM
To: marcin.wrobel
Subject: TriBITS Sphinx Documentation
When: Tuesday, July 13, 2021 9:00 AM-9:50 AM (UTC-05:00) Eastern Time (US & Canada).

Hello Marcin,

My Monday morning is pretty packed and I know Monday afternoon ET is no good for you on CET time so how about this time on Tuesday?

Some topics I would like to discuss at our meeting:

1.	Can we drive all generation from a fresh checkout of ‘master’ (or any other pristine branch or tag) with the Sphinx config.py file so there is no need for a temp branch to publish?
2.	Can we have a separate Sphinx documents for each of the three documents instead of one that includes all three?  (The whole purpose of working https://github.com/TriBITSPub/TriBITS/issues/381 to split off “TriBITS Users Guide” and “TriBITS Maintainers Guide” was so that users of TriBITS would not see any implementation details of TriBITS when looking at “TriBITS Users Guide”.)  Could we have a dummy top-level Sphinx document that just linked to these three documents but did not actually include them?  (We can discuss other ideas as well.)
3.	Can we use readthedocs.com for deployment?  What are the advantages and disadvantages of doing so?

Look forward to talking with you then.

-Ross

**Comment by marcinwrobel1986** [comment]
> FYI: I am setting up a meeting between myself and @marcinwrobel1986 (tentatively on Tuesday) to discuss above and related topics.
> 
> From: Bartlett, Roscoe A
> Sent: Sunday, July 11, 2021 6:46 PM
> To: marcin.wrobel
> Subject: TriBITS Sphinx Documentation
> When: Tuesday, July 13, 2021 9:00 AM-9:50 AM (UTC-05:00) Eastern Time (US & Canada).
> 
> Hello Marcin,
> 
> My Monday morning is pretty packed and I know Monday afternoon ET is no good for you on CET time so how about this time on Tuesday?
> 
> Some topics I would like to discuss at our meeting:
> 
> 1. Can we drive all generation from a fresh checkout of ‘master’ (or any other pristine branch or tag) with the Sphinx config.py file so there is no need for a temp branch to publish?
> 2. Can we have a separate Sphinx documents for each of the three documents instead of one that includes all three?  (The whole purpose of working [Split TribitsDevelopersGuide.* into TribitsUsersGuide.* and TribitsMaintainersGuide.* #381](https://github.com/TriBITSPub/TriBITS/issues/381) to split off “TriBITS Users Guide” and “TriBITS Maintainers Guide” was so that users of TriBITS would not see any implementation details of TriBITS when looking at “TriBITS Users Guide”.)  Could we have a dummy top-level Sphinx document that just linked to these three documents but did not actually include them?  (We can discuss other ideas as well.)
> 3. Can we use readthedocs.com for deployment?  What are the advantages and disadvantages of doing so?
> 
> Look forward to talking with you then.
> 
> -Ross

Hello Ross,
accepted. I will prepare the points you asked for. Thanks.

**Comment by marcinwrobel1986** [comment]
> @marcinwrobel1986, okay, I see the issue now. The issue is that each document is a single HTML page which shows up as "introduction" in as title in the search results so it is deceptive what is is showing you. Once you are on one of the three different document HTML pages, you can do "Ctrl+F". What throws you off is **Introduction** as the title for the search results and not the title of the document itself.
> 
> I appears there is no way to get Sphinx to automatically break up a long RST document into different HTML pages based on section headings with next and previous links a described in:
> 
> > https://stackoverflow.com/questions/44777009/can-i-split-a-single-restructuredtext-file-into-multiple-html-pages-by-section
> 
> The documentation at:
> 
> * https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html
> 
> says:
> 
> > Since reST does not have facilities to interconnect several documents, or split documents into multiple output files, Sphinx uses a custom directive to add relations between the single files the documentation is made of, as well as tables of contents. The toctree directive is the central element.
> 
> So this is a bit problematic. If you want really nice generated HTML documentation with each section getting its own HTML page, you can't use a single *.rst file and you have to use a Sphinx RST extension directive. That means if you use this Sphinx directive/extension, then you can no longer use simple docutils to generate HTML and PDF files but have to use Sphinx. But perhaps that is just the way it is. (It is just that I have had trouble in the past getting Sphinx to install on machines that I had to use. For example, I have not tried to install Sphinx on one of my RHEL 7 machines at SNL yet but I can imagine it will be a nightmare like all Python-based installs on RHEL 7. More on this in PR #387.)
> 

Ross, Sphinx is mainly used to generate automatic documentation with source code. As I mentioned before I haven't done anything with TriBITS documentation apart from changing linking, so Sphinx finds and put the file contetnt in correct places and the landing page as well. 
What the "new documentation" brings it:
- nice, well know theme
- with the theme, comes navigation and text formatting which is very readable
- possibility of automatic deployment to whatever server/service

> That means that the "Search docs" box is basically worthless. That is okay if you understand what is happening and that you just need to use "Ctrl+F" to search the HTML page but that was not immediately obvious to me. (And I realized that I might have mistyped 'tribits_add_test' in my earlier comment [above](https://github.com/TriBITSPub/TriBITS/issues/386#issuecomment-877225582) and that is why it did not hit anything with "Ctrl+F".)

It brings `phrase highlighting`, so always something in addition.

**Comment by marcinwrobel1986** [comment]
Hello Ross,

I have forced today the consequesces of what I wrote far above:

> currently the biggest concern is a fact that one needs to have python2 on the system which needs to be found under alias python (python2 not working), which consequence is that documentation can't be build when only python3 is available

The thing is that, Read The Docs is a service, which is trying to simplified documentation creation. It has some default behaviour, but there is a possibility to provide a configuration file to it. It works as application, you can not send a static web page there.
I wrote a config file, but there is a limitation to only one version of Python. It could be Python 2 or Python 3, but can not be both at the same time. 
E.g. I was able to configure Github Actions as there were both Python's interpreters present in the OS and there were no problems with the documentation at all.

Below is a screenshots that reveals the problem:
![Zrzut ekranu z 2021-07-13 11-28-02](https://user-images.githubusercontent.com/47597569/125439047-ef77b768-970d-4be0-9a78-62b2f14366de.png)

I was looking at PyPi for `Sphinx` and I have found, that the latest supported Sphinx version for Python2 was 2.1.1 (Current is 4.1.0 for Python3).

The error for generating documentation when Python2 is not present in the OS:
![Python2_error](https://user-images.githubusercontent.com/47597569/125439835-f6b2792b-b508-40a5-8ac2-531c78a01fdf.png)

So, the answer to a question `Could we deploy documentation to RTD?` would be, no we can not.
At the current state what could be done:
- GitHub pages, current solution, working
- deploy generated static documentation to a ftp or similar

Pros to stay with GitHub:
- repository is hosted on GitHub
- natural way to do things when hosted on GitHub
- free for OS projects
- custom domain could be used
- already setup

**Comment by bartlettroscoe** [comment]
> It could be Python 2 or Python 3, but can not be both at the same time.

@marcinwrobel1986, the TriBITS Python code being used here does not need Python 2, it works fine with Python 3.  There driver scripts might just need a small tweak to get that to work.  We can discuss.

**Comment by bartlettroscoe** [comment]
After discussing with @marcinwrobel1986 over MS Teams today related to PR #387, it seems like the best course of action to finish up this issue is:

* Break up into three separate Sphinx documents instead of one (using just a simple non Sphinx landing page that links to the three Sphinx documents)
* Continue using a GitHub Pages site with with a branch (called something like 'deploy-doc-site') and a GitHub Action to do the deployment.  (That gives the most flexibility and control.)
* Remove any need for Python 2 and DocUtils from the Sphinx documentation generation process.

The only question I have is if the documentation generation branch (currently called 'doc' in the repo https://github.com/marcinwrobel1986/TriBITS/) this should be in this main git repo http://github.com/TriBITSPub/TriBITS/ repo or in a separate deployment repo.  Committing generated files to the main repo (even if it is on a separate branch) just does not sit well with me (and others seem agree like [don't commit generated files](https://homes.cs.washington.edu/~mernst/advice/version-control.html#Dont_commit_generated_files), [do not commit generated files](https://acompiler.com/git-best-practices/), and [do not commit generated files](https://programmerfriend.com/git-best-practices/)).  I am just not excited to have a process that endlessly updates generated files on a branch in the main repo.  But then you seen arguments like [this](https://kentcdodds.com/blog/why-i-dont-commit-generated-files-to-master) that say that committing generated files to a special branch in the repo is okay.   So I guess I would need to see how big this separate branch (called something big and ugly like 'deploy-doc-site') would likely get and if it does not get too big, then it would be simpler to just keep that branch in the main git repo and move on.  (Simpler is better all things being equal.)

**Comment by bartlettroscoe** [comment]
@marcinwrobel1986, okay, now I see the `print "<string>"` error [above](https://github.com/TriBITSPub/TriBITS/issues/386#issuecomment-879055300) in the script `.py`.  Turns out that little script is not under any automated testing so it slipped through the cracks in the Python 3 upgrade of TriBITS.  (I will add an optional test that generates the TriBITS DocUtils documentation as one of the TriBITS tests and that will catch this going forward.)

**Comment by bartlettroscoe** [comment]
After merging #387, in this github repo's "Pages" settings, I sest:

![image](https://user-images.githubusercontent.com/1959736/126881350-95108e9b-537a-4388-9d32-4bd9307176f6.png)

and now you can see this site deployed at:

* http://tribitspub.github.io/TriBITS/

and the three documents are displayed there.  Nice!

**Comment by bartlettroscoe** [comment]
I updated the old [tribits.org site](https://tribits.org/) to also point to the new Sphinx-generated HTML documents.

Lastly, I will post a GitHub Issue for the remaining issues for the Sphinx-generated documentation before we can switch over and abandon the DocUtils-generated documentation.

**Comment by bartlettroscoe** [comment]
CC: @marcinwrobel1986 

To follow on from this, I created #399 to address what needs to be done to drop the DocUtils-generated documents and fully transition over to Sphinx with no regressions.  I also I created the follow-on issue #400 for taking full advantage of Sphinx after this transition to make these proper Sphinx/ReadTheDocs documents as people would expect.

NOTE: As part of the writing of #399 and a little more investigation, it seems that the formatting problems I am seeing with the Sphinx-generated documents is due to a problem with the `sphinx` and/or `sphinx-rtd-theme` installation on my SNL RHEL 7 machine (installed with `pip3 install --local sphinx sphinx-rtd-theme`).  When the Sphinx documentation is built with the GitHub Actions job, the formatting problem I saw is gone.  The problem was also gone when built using a sphinx and sphinx-rtd-theme installation on my Windows 10 laptop under Cygwin (using a Frankenstein setup mixing software installed for Cygwin and native Windows which is a miracle).  (It is issues like this why I gave up on Sphinx years ago when I tried.  It is just so time consuming fighting issues like this with complex software.)

I will now close this Issue and move on with the hard technical tasks in EPIC #367 where time is running out this FY.  Hopefully there will be time and $$ in FY22 to complete #399 and #400 and complete this transition to the Sphinx.

Thanks for all of your hard work on this!  What we have in place is pretty nice.  (Just a bit more work and I think we would be completely done.)
